### PR TITLE
Jacoco Coverage for Java

### PR DESCRIPTION
**Pull request by davidparry**
### **User description**
After accidentally closing #52 and than having to rework based on the other merge.
I have added the ability to run jacoco test coverage on a java project with Spockframework.

___

### **PR Type**
Enhancement, Tests, Documentation

___

### **Description**
- Added support for JaCoCo coverage reports in `CoverageProcessor`.
- Enhanced `UnitTestGenerator` to support Java and Groovy test insertion.
- Added new test cases for FastAPI endpoints.
- Added tests for JaCoCo report parsing in `CoverageProcessor`.
- Added end-to-end test for Java in `UnitTestGenerator`.
- Added README and Gradle wrapper scripts for Java Math application.
- Added `SimpleMathOperations` class and corresponding Spock tests.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement
</strong></td><td><details><summary>3 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>CoverageProcessor.py</strong><dd><code>Add support for JaCoCo coverage reports in CoverageProcessor.</code></dd></summary>
<hr>
      
cover_agent/CoverageProcessor.py

<li>Added support for JaCoCo coverage reports.<br> <li> Implemented methods to parse JaCoCo reports.<br> <li> Updated constructor to include <code>filepath</code>.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/cover-agent/pull/59/files#diff-610c623a5c322b67c9a12fa75021723cc335d4dc3ce4676eb8576a20f3ffd93e">+69/-3</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>UnitTestGenerator.py</strong><dd><code>Enhance UnitTestGenerator to support Java and Groovy tests.</code></dd></summary>
<hr>
      
cover_agent/UnitTestGenerator.py

<li>Updated <code>run_coverage</code> method to include <code>filepath</code>.<br> <li> Modified <code>validate_test</code> to handle Groovy and Java test insertion.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/cover-agent/pull/59/files#diff-19760582d9ede3a799fdbb541ad357b4822682e837bca8365196fba50daf57e3">+26/-6</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>SimpleMathOperations.java</strong><dd><code>Add SimpleMathOperations class with basic math operations.</code></dd></summary>
<hr>
      
templated_tests/java_gradle/src/main/java/com/davidparry/cover/SimpleMathOperations.java

- Added SimpleMathOperations class with basic math operations.

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/cover-agent/pull/59/files#diff-114616e77a5b5de574aa4c32439ddd8c3dee8064a049d88db56e9bdc5b803dce">+30/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Tests
</strong></td><td><details><summary>4 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>test_app.py</strong><dd><code>Add new test cases for FastAPI endpoints.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
templated_tests/python_fastapi/test_app.py

- Added new test cases for FastAPI endpoints.

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/cover-agent/pull/59/files#diff-474d2ef72228be13cfba725cbc72951c9dc719b9727c1c71a44ae20db4933fe7">+17/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>test_CoverageProcessor.py</strong><dd><code>Add tests for JaCoCo report parsing in CoverageProcessor.</code></dd></summary>
<hr>
      
tests/test_CoverageProcessor.py

<li>Added tests for JaCoCo report parsing.<br> <li> Updated existing tests to include <code>filepath</code>.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/cover-agent/pull/59/files#diff-63a3b2f65d74764f35622a99cfbd946c878c01230cf19bb3befc67d2dbfec0c6">+38/-1</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>test_UnitTestGenerator.py</strong><dd><code>Add end-to-end test for Java in UnitTestGenerator.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
tests/test_UnitTestGenerator.py

<li>Added end-to-end test for Java.<br> <li> Updated existing tests formatting.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/cover-agent/pull/59/files#diff-6563fc9db12b41fab8c40e7392e78047264cc44e6ed59a95ab1eccbecb84d8ad">+67/-8</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>SimpleMathOperationsSpec.groovy</strong><dd><code>Add Spock tests for SimpleMathOperations class.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
templated_tests/java_gradle/src/test/groovy/com/davidparry/cover/SimpleMathOperationsSpec.groovy

- Added Spock tests for SimpleMathOperations class.

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/cover-agent/pull/59/files#diff-b3f2b777ec797e789ce838f5f76103ae3eb1e6ea24f0f8b58443294d507ba3c4">+28/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Documentation
</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>README.md</strong><dd><code>Add README for Java Math application.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
templated_tests/java_gradle/README.md

- Added README for Java Math application.

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/cover-agent/pull/59/files#diff-5e937021f54b17f1ee443927c052741f6c7e1f859a03a9268be0a3578038cc85">+17/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Configuration changes
</strong></td><td><details><summary>3 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>gradle-wrapper.properties</strong><dd><code>Add Gradle wrapper properties for Java project.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
templated_tests/java_gradle/gradle/wrapper/gradle-wrapper.properties

- Added Gradle wrapper properties for Java project.

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/cover-agent/pull/59/files#diff-941cb993d9d05364c5a5d6e58b57144de9518e8d77f98dd07c9ce4233dfdba16">+6/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>gradlew</strong><dd><code>Add Gradle wrapper script for Unix systems.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
templated_tests/java_gradle/gradlew

- Added Gradle wrapper script for Unix systems.

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/cover-agent/pull/59/files#diff-7d2e3d60a5834ca94f4bb9c8f286c0df5b0efa5135a7fa57c6fccbdff097f615">+234/-0</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>gradlew.bat</strong><dd><code>Add Gradle wrapper script for Windows systems.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
templated_tests/java_gradle/gradlew.bat

- Added Gradle wrapper script for Windows systems.

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/cover-agent/pull/59/files#diff-c2cac7edf84ed44bbdc946efed0f71bc672b7aa84af47d27cdfcc996abb79619">+89/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

**Comment by davidparry** [comment]
> _PR-Agent was enabled for this repository. To continue using it, please link your git user with your CodiumAI identity [here](https://app.codium.ai/pr-agent)._
> 
> ## PR Code Suggestions ✨
> Category	Suggestion                                                                                                                                   	Score
> **Possible issue**	
> Correct the format of Groovy test code in the `CANNED_TESTS` dictionary
> 10
> **Possible bug**	
> Handle out-of-bounds `relevant_line_number_to_insert_after` in `validate_test`
> **The `validate_test` method should handle the case where `relevant_line_number_to_insert_after` is out of bounds, which could cause an IndexError.**
> 
> [cover_agent/UnitTestGenerator.py [316]](https://github.com/Codium-ai/cover-agent/pull/59/files#diff-19760582d9ede3a799fdbb541ad357b4822682e837bca8365196fba50daf57e3R316-R316)
> 
> ```diff
> +if relevant_line_number_to_insert_after > len(original_content_lines):
> +    relevant_line_number_to_insert_after = len(original_content_lines)
>  processed_test_lines = original_content_lines[:relevant_line_number_to_insert_after] + test_code_lines + original_content_lines[relevant_line_number_to_insert_after:]
>  
> ```
> 
> Suggestion importance[1-10]: 8
> Why: This suggestion correctly identifies a potential bug that could cause an IndexError and provides a solution to handle it, which is crucial for robustness.
> 
> 8
> Correct the path separators for the `JAVA_EXE` variable to ensure compatibility with Windows paths
> 6
> **Performance**	
> Replace the recursive Fibonacci implementation with an iterative approach to improve performance and avoid stack overflow
> 8
> **Enhancement**	
> Return actual covered and missed lines instead of empty lists in `parse_coverage_report_jacoco`
> 7
> **Robustness**	
> Add error handling for `ls` and `cd` commands to make the script more robust
> 7
> **Test coverage**	
> Add a test case for the `divide` method to check for division by zero and normal division
> 7
> **Best practice**	
> Add a check to ensure the response JSON contains the expected key in `test_subtract`
> 6

The possible bug is part of the javascript merge. the other related comments are about the example project which is not relevant its just an example to add tests.

**Comment by davidparry** [comment]
> _PR-Agent was enabled for this repository. To continue using it, please link your git user with your CodiumAI identity [here](https://app.codium.ai/pr-agent)._
> 
> ## PR Code Suggestions ✨
> Category	Suggestion                                                                                                                                   	Score
> **Possible issue**	
> Correct the format of Groovy test code in the `CANNED_TESTS` dictionary
> 10
> **Possible bug**	
> Handle out-of-bounds `relevant_line_number_to_insert_after` in `validate_test`
> 8
> Correct the path separators for the `JAVA_EXE` variable to ensure compatibility with Windows paths
> 6
> **Performance**	
> Replace the recursive Fibonacci implementation with an iterative approach to improve performance and avoid stack overflow
> 8
> **Enhancement**	
> Return actual covered and missed lines instead of empty lists in `parse_coverage_report_jacoco`
> 7
> **Robustness**	
> Add error handling for `ls` and `cd` commands to make the script more robust
> 7
> **Test coverage**	
> Add a test case for the `divide` method to check for division by zero and normal division
> 7
> **Best practice**	
> Add a check to ensure the response JSON contains the expected key in `test_subtract`
> 6

**Addressing the Design Concern:** This is a larger effort to refactor the codebase. First, the refactor should happen without introducing any new functionality. A new pattern should emerge to deal with different coverage frameworks, possibly using a factory pattern based on the `--coverage-type` parameter, which would trigger the appropriate classes.

Given my background in Java, I would suggest making `UnitTestGenerator` and `CoverageProcessor` base classes or interfaces, as they likely have common code to share with different classes inheriting them. Common tasks include inserting the suggested new tests into specific unit tests and working with the output of different types of coverage products. Just my opinion.

**Comment by davidparry** [comment]
> Error Handling:

On the Error Handling: for extract_package_and_class the FileNotFoundError error will occur before reaching this point during the testing of the source file in parameters. there is not way for filepath to not point to a file since the test fails and exits the program.

File: cover_agent/CoverageProcessor.py [hunk 1]
```diff
@@ -116,3 +122,63 @@ def parse_coverage_report_cobertura(self) -> Tuple[list, list, float]:
...
+        I expect this should bring up a discussion on introduce a factory for different CoverageProcessors. Where the
+        total coverage percentage is returned to be evaluated only.
+        """
+        lines_covered, lines_missed = [], []
+
+        package_name, class_name = self.extract_package_and_class()
+        missed, covered = self.parse_missed_covered_lines(package_name, class_name)
+
+        total_lines = missed + covered
+        coverage_percentage = (covered / total_lines) if total_lines > 0 else 0
```
Comment (davidparry) [inline cover_agent/CoverageProcessor.py:143]: I am following the pattern established from line 121 in this file for cobertura, also the UnitTestGenerator.py line 150 actually does the multiplication so the report would be wrong for the report.

File: templated_tests/python_fastapi/test_app.py [hunk 2]
```diff
@@ -11,3 +11,20 @@ def test_root():
...
+    response = client.get("/add/2/3")
+    assert response.status_code == 200
+    assert "result" in response.json()
+    assert response.json()["result"] == 5
+
+def test_subtract():
+    response = client.get("/subtract/5/2")
+    assert response.status_code == 200
+    assert "result" in response.json()
+    assert response.json()["result"] == 3
```
Comment (EmbeddedDevops1) [inline templated_tests/python_fastapi/test_app.py:30]: Remove generated tests.
```suggestion
```

File: cover_agent/CoverageProcessor.py [hunk 3]
```diff
@@ -116,3 +122,63 @@ def parse_coverage_report_cobertura(self) -> Tuple[list, list, float]:
...
+
+        return lines_covered, lines_missed, coverage_percentage
+
+    def parse_missed_covered_lines(self, package_name: str, class_name: str) -> tuple[int, int]:
```
Comment (EmbeddedDevops1) [inline cover_agent/CoverageProcessor.py:147]: Is this `jacoco` specific? If so, it may be prudent to append `jacoco` to the end of the function name. 

If we start to support more coverage types we may want to move coverage processors to a dedicated folder and have coverage types as their own class that inherit and override functions from a base class.

File: cover_agent/CoverageProcessor.py [hunk 4]
```diff
@@ -116,3 +122,63 @@ def parse_coverage_report_cobertura(self) -> Tuple[list, list, float]:
...
+            missed, covered = 0, 0
+            for row in reader:
+                if row['PACKAGE'] == package_name and row['CLASS'] == class_name:
+                    missed = int(row['LINE_MISSED'])
+                    covered = int(row['LINE_COVERED'])
+                    break
+
+        return missed, covered
+
+    def extract_package_and_class(self):
```
Comment (EmbeddedDevops1) [inline cover_agent/CoverageProcessor.py:159]: Same question as [above](https://github.com/Codium-ai/cover-agent/pull/59/files#r1618118458).

File: cover_agent/CoverageProcessor.py [hunk 5]
```diff
@@ -28,11 +31,12 @@ def __init__(
         """
         self.file_path = file_path
         self.filename = filename
+        self.filepath = filepath
```
Comment (EmbeddedDevops1) [inline cover_agent/CoverageProcessor.py:34]: Any specific reason why you're using `filepath` versus the already declared `file_path`? Feel free to refactor with or without the `_` if you think it helps with readability.

File: cover_agent/UnitTestGenerator.py [hunk 6]
```diff
@@ -292,14 +293,32 @@ def validate_test(self, generated_test: dict, generated_tests_dict: dict):
...
+                            + original_content[last_brace_index:]
+                    )
+            else:
+                # insert the test code at the relevant line
+                processed_test_lines = original_content_lines[:relevant_line_number_to_insert_after] + test_code_lines + original_content_lines[relevant_line_number_to_insert_after:]
+                processed_test = "\n".join(processed_test_lines)
+                # insert the additional imports at the top of the file
+                if additional_imports and additional_imports.rstrip() not in processed_test:
+                    processed_test = additional_imports.rstrip() + "\n\n" + processed_test
+
```
Comment (EmbeddedDevops1) [inline cover_agent/UnitTestGenerator.py:321]: @mrT23 Is this something you had planned on implementing (or implemented)? Specifically the indented test code...

**Comment by EmbeddedDevops1** [comment]
@davidparry First off, thank you very much for spending the time to get `jacoco` coverage parsing supported for this community project. Really appreciate it.

We've made some pretty major changes recently and it looks like you'll need to rebase (`git pull origin main --rebase`) and resolve conflicts locally and then push your changes (`git push -f`). Please have a look when you get a chance. Thanks again.

**Comment by davidparry** [comment]
> @davidparry First off, thank you very much for spending the time to get `jacoco` coverage parsing supported for this community project. Really appreciate it.
> 
> We've made some pretty major changes recently and it looks like you'll need to rebase (`git pull origin main --rebase`) and resolve conflicts locally and then push your changes (`git push -f`). Please have a look when you get a chance. Thanks again.

I am not sure how the my PR was closed again by me i did make updates and created a fresh PR https://github.com/Codium-ai/cover-agent/pull/62

File: cover_agent/UnitTestGenerator.py (hunk 6 above)
Comment (davidparry) [reply cover_agent/UnitTestGenerator.py]: after your refactor and it appears the new way you are inserting into the test file it is not no longer needed.
